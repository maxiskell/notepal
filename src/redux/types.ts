export interface INote {
  id: number;
  title: string;
  content: string;
}

export interface INoteState {
  notes: Array<INote>;
  editorOpen: boolean;
  currentNoteId: number | null;
}
