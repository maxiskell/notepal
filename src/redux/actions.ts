import { INote } from "./types";

export enum NoteActionTypes {
  ADD = "ADD_NOTE",
  UPDATE = "UPDATE",
  OPEN_EDITOR = "OPEN_EDITOR",
  CLOSE_EDITOR = "CLOSE_EDITOR",
}

export interface IAddNote {
  type: NoteActionTypes.ADD;
  payload: INote;
}

export interface IUpdateNote {
  type: NoteActionTypes.UPDATE;
  payload: INote;
}

export interface IOpenNoteEditor {
  type: NoteActionTypes.OPEN_EDITOR;
  // if no note given, it means "new note"
  payload: string | null;
}

export interface ICloseNoteEditor {
  type: NoteActionTypes.CLOSE_EDITOR;
}

export type NoteAction =
  | IAddNote
  | IUpdateNote
  | IOpenNoteEditor
  | ICloseNoteEditor;

export const openEditor = (noteId: string | null): NoteAction => ({
  type: NoteActionTypes.OPEN_EDITOR,
  payload: noteId,
});

export const closeEditor = (): NoteAction => ({
  type: NoteActionTypes.CLOSE_EDITOR,
});

export const addNote = (note: INote): NoteAction => ({
  type: NoteActionTypes.ADD,
  payload: note,
});

export const updateNote = (note: INote): NoteAction => ({
  type: NoteActionTypes.UPDATE,
  payload: note,
});
