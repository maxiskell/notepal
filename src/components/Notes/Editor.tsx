import React, { useState, ChangeEvent } from "react";

import { NoteAction } from "../../redux/actions";
import { INote } from "../../redux/types";

interface IEditorProps {
  note?: INote;
  save: (note: INote) => NoteAction;
  close: (event: React.MouseEvent<HTMLButtonElement>) => NoteAction;
}

const Editor: React.FC<IEditorProps> = (props) => {
  const [title, setTitle] = useState(props.note?.title || "");
  const [content, setContent] = useState(props.note?.content || "");

  return (
    <div className="container mt-4" data-testid="note-editor">
      <div className="form-group">
        <label htmlFor="title">Title</label>
        <input
          required
          id="title"
          type="text"
          name="title"
          className="form-control"
          data-testid="note-editor-title"
          value={title}
          onChange={(e: ChangeEvent<HTMLInputElement>) =>
            setTitle(e.target.value)
          }
        />
      </div>
      <div className="form-group">
        <label htmlFor="title">Content</label>
        <textarea
          required
          rows={10}
          id="content"
          name="content"
          className="form-control"
          data-testid="note-editor-content"
          value={content}
          onChange={(e: ChangeEvent<HTMLTextAreaElement>) =>
            setContent(e.target.value)
          }
        />
      </div>

      <div className="btn-group float-right">
        <button
          type="reset"
          className="btn btn-secondary"
          data-testid="note-editor-cancel"
          onClick={props.close}
        >
          Cancel
        </button>
        <button
          type="submit"
          className="btn btn-primary"
          data-testid="note-editor-save"
          onClick={() => props.save({ id: props.note?.id, title, content })}
        >
          Save
        </button>
      </div>
    </div>
  );
};

export default Editor;
