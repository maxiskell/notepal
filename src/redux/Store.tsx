import { combineReducers, createStore, Store } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";

import { INoteState } from "./types";
import { noteReducer } from "./reducer";

export interface IAppState {
  noteState: INoteState;
}

const rootReducer = combineReducers<IAppState>({
  noteState: noteReducer,
});

export default function configureStore(): Store<IAppState, any> {
  const store = createStore(rootReducer, composeWithDevTools());
  return store;
}
