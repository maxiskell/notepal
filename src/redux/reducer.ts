import { Reducer } from "redux";
import { v4 as uuid } from "uuid";

import { INoteState, INote } from "./types";
import { NoteAction, NoteActionTypes } from "./actions";

const initialState: INoteState = {
  notes: new Map() as Map<string, INote>,
  editorOpen: false,
  currentNoteId: null,
};

export const noteReducer: Reducer<INoteState, NoteAction> = (
  state = initialState,
  action
) => {
  let notes = state.notes;

  switch (action.type) {
    case NoteActionTypes.ADD:
      notes.set(uuid(), action.payload);

      return {
        ...state,
        editorOpen: false,
        notes,
      };
    case NoteActionTypes.UPDATE:
      notes.set(state.currentNoteId!, action.payload);

      return {
        ...state,
        notes,
        editorOpen: false,
      };
    case NoteActionTypes.OPEN_EDITOR:
      return {
        ...state,
        editorOpen: true,
        currentNoteId: action.payload,
      };
    case NoteActionTypes.CLOSE_EDITOR:
      return {
        ...state,
        editorOpen: false,
        currentNoteId: null,
      };
    default:
      return state;
  }
};
