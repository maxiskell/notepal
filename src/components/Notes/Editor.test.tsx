import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Editor from "./Editor";
import { closeEditor } from "../../redux/actions";

test("renders properly", () => {
  const { container } = render(<Editor close={closeEditor} />);
  expect(container).toMatchSnapshot();
});

test("cancel button calls close function", async () => {
  const mockClose = jest.fn();
  const { getByText } = render(<Editor close={mockClose} />);
  fireEvent.click(getByText(/Cancel/i));
  expect(mockClose).toHaveBeenCalledTimes(1);
});
