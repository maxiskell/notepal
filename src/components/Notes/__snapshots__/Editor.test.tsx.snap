// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div>
  <div
    class="mt-4"
  >
    <div
      class="form-group"
    >
      <label
        for="title"
      >
        Title
      </label>
      <input
        class="form-control"
        id="title"
        name="title"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="title"
      >
        Content
      </label>
      <textarea
        class="form-control"
        id="content"
        name="content"
        required=""
        rows="10"
      />
    </div>
    <div
      class="btn-group float-right"
    >
      <button
        class="btn btn-secondary"
        type="reset"
      >
        Cancel
      </button>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
